MaxText/maxtext_utils.py

```diff
@@ -0,0 +1,109 @@
+"""
+ Copyright 2023 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+# pylint: disable=bare-except, consider-using-generator
+"""Utils that are only interesting to MaxText. """
+
+import jax
+from jax.sharding import PartitionSpec as P
+from jax.experimental.serialize_executable import deserialize_and_load
+import pickle
+import functools
+import input_pipeline
+import optax
+import max_utils
+
+def get_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config):
+  """ Get the shardings (both state and data) for train_step """
+  functional_train = get_functional_train_step(train_step, model, config)
+  data_pspec = P(*config.data_sharding)
+  state_mesh_shardings = jax.tree_map(
+      lambda p: jax.sharding.NamedSharding(mesh, p), state_mesh_annotations)
+  data_sharding = jax.tree_map(
+      lambda p: jax.sharding.NamedSharding(mesh, p), data_pspec)
+  in_shardings = (state_mesh_shardings, data_sharding, None) # State, batch, rng
+  out_shardings = (state_mesh_shardings, None, None) # State, metrics, rng
+  static_argnums = () # We partial out the static argnums of model and config
+  donate_argnums = 0 # This is the index of the state - we allow the compiler to make use of this memory.
+  return functional_train, in_shardings, out_shardings, static_argnums, donate_argnums
+
+def get_functional_train_step(train_step, model, config):
+  return functools.partial(train_step, model, config)
+
+def get_optimizer(config, learning_rate_schedule):
+  """ Create AdamW Optimizer following Llama2's training details, see https://arxiv.org/pdf/2307.09288.pdf section 2.2 """
+  return optax.adamw(
+    learning_rate_schedule,
+    b1=config.adam_b1,
+    b2=config.adam_b2,
+    eps=config.adam_eps,
+    eps_root=config.adam_eps_root,
+    weight_decay=config.adam_weight_decay,
+  )
+
+def validate_train_config(config):
```

Comment:
Wait shouldn't this be in train.py? It is specific to train.py?

Reply:
Moved to train.py